File: app/src/main/java/ru/geekbrains/notes/observation/Publisher.java
package ru.geekbrains.notes.observation;

import java.util.ArrayList;
import java.util.List;

import ru.geekbrains.notes.data.CardData;

public class Publisher {
    private List<Observer> observers;

    public Publisher() {
        this.observers = new ArrayList<Observer>();
    }

    public void subscribe(Observer observer) {
        observers.add(observer);
    }

    public void unsubscribe(Observer observer) {
        observers.remove(observer);
    }

    public void notifyTask(CardData cardData) {
        for (Observer observer : observers) {
            observer.updateState(cardData);
            unsubscribe(observer);
        }
    }
}

File: app/src/main/java/ru/geekbrains/notes/observation/Publisher.kt
package ru.geekbrains.notes.observation

import ru.geekbrains.notes.model.CardData

class Publisher {
    private val observers: MutableList<Observer>
    fun subscribe(observer: Observer) {
        observers.add(observer)
    }

    private fun unsubscribe(observer: Observer) {
        observers.remove(observer)
    }

    fun notifyTask(cardData: CardData) {
        for (observer in observers) {
            observer.updateState(cardData)
            unsubscribe(observer)
        }
    }

    init {
        observers = ArrayList()
    }
}
